### PR TITLE
test_cpu and test_gpu EvalModelTemplate deprecation

### PR DESCRIPTION
**Pull request by gianscarpe**
## What does this PR do?
This PR deprecates EvalModelTemplate and old pl APIs in favor of BoringModel in `tests/models/test_cpu` and `tests/models/test_gpu` . BoringModel is inherited in for `test_tbptt` in `test_cpu`. A dispatcher function guarantees `EvalModelTemplate` backward compatibility while completing the deprecation 

Fixes # (issue)

## Before submitting
- [X] Was this discussed/approved via a Github issue? (no need for typos and docs improvements)
- [X] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), Pull Request section?
- [X] Did you make sure your PR does only one thing, instead of bundling different changes together? Otherwise, we ask you to create a separate PR for every change.
- [X] Did you make sure to update the documentation with your changes?
- [X] Did you write any new necessary tests? 
- [X] Did you verify new and existing tests pass locally with your changes?
- [X] If you made a notable change (that affects users), did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)?

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:  

 - [X] Is this pull request ready for review? (if not, please submit in draft mode)
 - [X] Check that all items from **Before submitting** are resolved
 - [X] Make sure the title is self-explanatory and the description concisely explains the PR
 - [X] Add labels and milestones (and optionally projects) to the PR so it can be classified; _Bugfixes should be including in bug-fix release milestones (m.f.X) and features should be included in (m.X.b) releases._
 

## Did you have fun?
Absolutely!

### CODE REVIEW

**Comment by gianscarpe** [comment]
First PR for refactoring tests and deprecating EvalModelTemplater is ready for review! :100:  @SeanNaren

File: tests/base/boring_model.py [hunk 1]
```diff
@@ -55,8 +56,14 @@ def __len__(self):
 
 
 class BoringModel(LightningModule):
-
-    def __init__(self):
+    def __init__(
```
Comment (Borda) [inline tests/base/boring_model.py:59]: let's keep the boring model simple if needed inherit and add extra argiments

File: tests/base/boring_model.py [hunk 2]
```diff
@@ -75,7 +82,12 @@ def training_step(...):
 
         """
         super().__init__()
-        self.layer = torch.nn.Linear(32, 2)
+        self.layer = torch.nn.Linear(in_features, out_features)
+        self.batch_size = batch_size
```
Comment (Borda) [inline tests/base/boring_model.py:86]: we shall enforce the laters API, so use `save_hyperparameters()`

File: tests/base/boring_model.py [hunk 3]
```diff
@@ -117,15 +126,26 @@ def test_epoch_end(self, outputs) -> None:
         torch.stack([x["y"] for x in outputs]).mean()
 
     def configure_optimizers(self):
-        optimizer = torch.optim.SGD(self.layer.parameters(), lr=0.1)
+        optimizer = getattr(torch.optim, self.optimizer_name)(self.layer.parameters(), lr=self.learning_rate)
```
Comment (Borda) [inline tests/base/boring_model.py:129]: lets split it in two lines

File: tests/base/boring_model.py [hunk 4]
```diff
@@ -117,15 +126,26 @@ def test_epoch_end(self, outputs) -> None:
...
 
     def val_dataloader(self):
-        return torch.utils.data.DataLoader(RandomDataset(32, 64))
+        return torch.utils.data.DataLoader(RandomDataset(32, 64), batch_size=self.batch_size)
 
     def test_dataloader(self):
-        return torch.utils.data.DataLoader(RandomDataset(32, 64))
+        return torch.utils.data.DataLoader(RandomDataset(32, 64), batch_size=self.batch_size)
+
+    def get_default_hparams(continue_training: bool = False, hpc_exp_number: int = 0) -> dict:
```
Comment (Borda) [inline tests/base/boring_model.py:142]: this is out f the scope of this simple model, any way you request some arguments which are not later used

File: tests/base/develop_pipelines.py [hunk 5]
```diff
@@ -117,3 +126,21 @@ def run_prediction(dataloader, trained_model, dp=False, min_acc=0.50):
         acc = acc.item()
 
     assert acc >= min_acc, f"This model is expected to get > {min_acc} in test set (it got {acc})"
+
+
+@run_prediction.register(BoringModel)
```
Comment (Borda) [inline tests/base/develop_pipelines.py:131]: this makes the simple model complicated, lest move all these to another model

**Review by Borda — CHANGES_REQUESTED** [review]

**Comment by rohitgr7** [comment]
@gianscarpe it's not required to update the branch every time it gets out-of-date with the base :]

File: tests/base/boring_model.py (hunk 1 above)
Comment (gianscarpe) [reply tests/base/boring_model.py:59]: Thank you for the review! Do we keep instantiation without arguments?

File: tests/base/develop_pipelines.py (hunk 5 above)
Comment (gianscarpe) [reply tests/base/develop_pipelines.py:131]: I agree! We can add different dispatcher for more complicated models (e.g., models specifically for dp testing)

File: tests/base/boring_model.py [hunk 6]
```diff
@@ -75,7 +82,14 @@ def training_step(...):
...
         super().__init__()
-        self.layer = torch.nn.Linear(32, 2)
+        self.layer = torch.nn.Linear(in_features, out_features)
+        self.batch_size = batch_size
+        self.in_features = in_features
+        self.learning_rate = learning_rate
+        self.optimizer_name = optimizer_name
+        self.out_features = out_features
+        self.save_hyperparameters('batch_size', 'in_features', 'out_features',
+                                  'optimizer_name', 'learning_rate')
```
Comment (rohitgr7) [inline tests/base/boring_model.py:92]:
```suggestion
        self.save_hyperparameters()
```
you can just use `self.save_hyperparameters()` and access them using `self.hparams.argument`.

Comment (gianscarpe) [reply tests/base/boring_model.py:92]: Hi @rohitgr7, I have a question regarding `save_hyperparameters()`. In particular, if I call it in __init__ method of `BoringModel`, than subclass model `RuntimeParamChangeModelAssign` in `test_hparams.py` fail to save their hypermaremeters (specifically if args are assigned to `self.hparams` as in line 601 in 'tests/models/test_hparams.py')

Comment (rohitgr7) [reply tests/base/boring_model.py]: assigning `hparams` like `self.hparams = params` is not the recommended way and was deprecated due to this reason only AFAIR. use `self.save_hyperparameters()` there too.

Comment (rohitgr7) [reply tests/base/boring_model.py]:
```python
class RuntimeParamChangeModelAssign(BoringModel):
    def __init__(self, **kwargs):
        super().__init__()
        self.save_hyperparameters()
```

Comment (awaelchli) [reply tests/base/boring_model.py]: Just as a side note for completeness sake:
It is totally fine to assign `self.something = something`. The only exception is self.hparams, that was an attribute with special behaviour and this is now deprecated. save_hyperparameters() will bundle all args into a namespace/dict accessible through self.hparams. It is up to the user to decide if they want to read from it or not.

both the current changes and the ones proposed by @rohitgr7 look good to me :)

Comment (gianscarpe) [reply tests/base/boring_model.py]: Thank you both for the explanations! :) :top:

File: tests/base/boring_model.py [hunk 7]
```diff
@@ -75,7 +75,12 @@ def training_step(...):
 
         """
         super().__init__()
-        self.layer = torch.nn.Linear(32, 2)
+
+        self.batch_size = 1
+        self.learning_rate = 0.1
+        self.optimizer_name = "SGD"
```
Comment (rohitgr7) [inline tests/base/boring_model.py:81]: why not keep this in hparams??

File: tests/base/boring_model.py [hunk 8]
```diff
@@ -117,15 +119,16 @@ def test_epoch_end(self, outputs) -> None:
         torch.stack([x["y"] for x in outputs]).mean()
 
     def configure_optimizers(self):
-        optimizer = torch.optim.SGD(self.layer.parameters(), lr=0.1)
+        optimizer_class = getattr(torch.optim, self.optimizer_name)
+        optimizer = optimizer_class(self.layer.parameters(), lr=self.learning_rate)
```
Comment (rohitgr7) [inline tests/base/boring_model.py:123]:
```suggestion
        optimizer = optimizer_class(self.parameters(), lr=self.learning_rate)
```

File: tests/base/develop_pipelines.py [hunk 9]
```diff
@@ -39,59 +43,64 @@ def run_model_test_without_loggers(trainer_options, model, min_acc: float = 0.50
...
     reset_seed()
-    save_dir = trainer_options['default_root_dir']
+    save_dir = trainer_options["default_root_dir"]
 
     # logger file to get meta
     logger = get_default_logger(save_dir, version=version)
     trainer_options.update(logger=logger)
 
-    if 'checkpoint_callback' not in trainer_options:
+    # DEPRECATED option
```
Comment (rohitgr7) [inline tests/base/develop_pipelines.py:64]:
```suggestion
    # TODO: DEPRECATED option
```
just for visibility to make sure to update this when it gets removed.

File: tests/base/develop_pipelines.py [hunk 10]
```diff
@@ -117,3 +126,14 @@ def run_prediction(dataloader, trained_model, dp=False, min_acc=0.50):
...
+
+@run_prediction.register(BoringModel)
+def _(trained_model, dataloader, dp=False, min_acc=0.25):
+    # run prediction on 1 batch
+    batch = next(iter(dataloader))
+    with torch.no_grad():
+        output = trained_model(batch)
+    acc = trained_model.loss(batch, output)
+
+    assert acc >= min_acc, f"This model is expected to get , {min_acc} in test set (it got {acc})"
```
Comment (rohitgr7) [inline tests/base/develop_pipelines.py:139]:
```suggestion
    assert acc >= min_acc, f"This model is expected to get, {min_acc} in test set but got {acc}"
```

File: tests/base/develop_pipelines.py [hunk 11]
```diff
@@ -117,3 +126,14 @@ def run_prediction(dataloader, trained_model, dp=False, min_acc=0.50):
         acc = acc.item()
 
     assert acc >= min_acc, f"This model is expected to get > {min_acc} in test set (it got {acc})"
+
+
+@run_prediction.register(BoringModel)
+def _(trained_model, dataloader, dp=False, min_acc=0.25):
+    # run prediction on 1 batch
+    batch = next(iter(dataloader))
+    with torch.no_grad():
```
Comment (rohitgr7) [inline tests/base/develop_pipelines.py:135]: you can put this as a decorator. `@torch.no_grad()`.

File: tests/models/test_cpu.py [hunk 12]
```diff
@@ -75,57 +70,85 @@ def test_cpu_slurm_save_load(tmpdir):
     # new logger file to get meta
     logger = tutils.get_default_logger(tmpdir, version=version)
 
+    model = BoringModel()
+
+    class _StartCallback(Callback):
+        def on_init_start(self, trainer):
+            print("Starting to init trainer!")
+
```
Comment (rohitgr7) [inline tests/models/test_cpu.py:78]:
```suggestion
```
is this required?

File: tests/models/test_cpu.py [hunk 13]
```diff
@@ -75,57 +70,85 @@ def test_cpu_slurm_save_load(tmpdir):
...
-
     # by calling fit again, we trigger training, loading weights from the cluster
     # and our hook to predict using current model before any more weight updates
     trainer.fit(model)
 
 
-def test_early_stopping_cpu_model(tmpdir):
+def test_all_features_cpu_model(tmpdir):
     """Test each of the trainer options."""
-    stopping = EarlyStopping(monitor='early_stop_on', min_delta=0.1)
```
Comment (rohitgr7) [inline tests/models/test_cpu.py:104]: I think you need to add the `EarlyStopping` callback back here since it's not added by default. Merge the below test here itself?

File: tests/models/test_cpu.py [hunk 14]
```diff
@@ -186,7 +212,29 @@ def test_default_logger_callbacks_cpu_model(tmpdir):
...
+            self.log('val_loss', val_loss)
+
+        def test_step(self, batch, batch_idx):
+            output = self.layer(batch)
+            loss = self.loss(batch, output)
+            return {"y": loss}
+
+        def test_epoch_end(self, outputs) -> None:
+            test_loss = torch.stack([x["y"] for x in outputs]).mean()
+            self.log('test_loss', test_loss)
```
Comment (rohitgr7) [inline tests/models/test_cpu.py:235]:
```suggestion
        def validation_step(self, batch, batch_idx):
            output = self.layer(batch)
            loss = self.loss(batch, output)
            self.log('val_loss', loss)

        def test_step(self, batch, batch_idx):
            output = self.layer(batch)
            loss = self.loss(batch, output)
            self.log('test_loss', loss)
```
let lightning aggregate the loss over the batches.

File: tests/models/test_cpu.py [hunk 15]
```diff
@@ -202,22 +250,38 @@ def test_running_test_after_fitting(tmpdir):
...
+            pass
+
+        def test_step(self, batch, batch_idx):
+            output = self.layer(batch)
+            loss = self.loss(batch, output)
+            return {"y": loss}
+
+        def test_epoch_end(self, outputs) -> None:
+            test_loss = torch.stack([x["y"] for x in outputs]).mean()
+            self.log('test_loss', test_loss)
```
Comment (rohitgr7) [inline tests/models/test_cpu.py:282]:
```suggestion
        def test_step(self, batch, batch_idx):
            output = self.layer(batch)
            loss = self.loss(batch, output)
            self.log('test_loss', loss)
```
same here.

File: tests/models/test_cpu.py [hunk 16]
```diff
@@ -75,57 +70,85 @@ def test_cpu_slurm_save_load(tmpdir):
...
+        def on_init_start(self, trainer):
+            print("Starting to init trainer!")
+
+        # set the epoch start hook so we can predict before the model does the full training
+        def on_epoch_start(self, trainer, model):
+            assert trainer.global_step == real_global_step and trainer.global_step > 0
+            # predict with loaded model to make sure answers are the same
+            model.eval()
+            new_pred = model(batch)
+            assert torch.all(torch.eq(pred_before_saving, new_pred)).item() == 1
```
Comment (rohitgr7) [inline tests/models/test_cpu.py:85]:
```suggestion
            mode = model.training
            model.eval()
            new_pred = model(batch)
            assert torch.eq(pred_before_saving, new_pred).all()
            model.train(mode)
```
should be put back to the training mode.

File: tests/models/test_cpu.py (hunk 15 above)
Comment (gianscarpe) [reply tests/models/test_cpu.py:284]: I discussed with @Borda about BoringModel, and we decided to keep the model as minimal as possible. Specifically, if a test need an additional functionality (e.g., set the learning rate, set the optimizer, aggregate the return values and epoch_end), it should inherit the basic BoringModel in a subclass that implements these features. For examples, logging is tested in `test_hparams`, optimizers and lr are tested in `test_cpu`, and so on. This also grantee more compatibility with exisiting tests. If I misinterpreted @Borda 's suggestion, I can fix these!

File: tests/models/test_cpu.py (hunk 13 above)
Comment (gianscarpe) [reply tests/models/test_cpu.py:104]: Hi @rohitgr7, thanks for the suggestion! However, I organized the code to tests callbacks separately from other trainer options to keep tests simpler and cleaner. Let me know if it sounds good :)

Comment (rohitgr7) [reply tests/models/test_cpu.py:104]: All the callbacks are already tested in tests/callbacks. This test doesn't do anything special, just takes all the arguments even the ModelCheckpoint will be added by default. Only EarlyStopping is missing, so I think it's not necessary here to create a separate one.

File: tests/models/test_cpu.py (hunk 15 above)
Comment (rohitgr7) [reply tests/models/test_cpu.py:281]: If BordingModel should be kept minimal, I'd suggest removing `self.optimizer`, `self.batch_size` and `self.learning_rate` too. If required one should override these methods wherever it's required.

Regarding the methods, `self.log()` with `on_epoch=True`(default) in `test_step` will aggregate the results for you. No need to override `epoch_end`. Same for validation_step.

Comment (gianscarpe) [reply tests/models/test_cpu.py:281]: Regarding the parameters, I proposed a solution that doesn't require input parameters, while still defining parameters instead of using them as "magic constants". I believe this solution should be clean enough, maybe @Borda can help us on this matter.

On the other hand, `self.log()` introduce a new behavior to `BoringModel` that wasn't conceived, and therefore if I add the logging of the aggregation results other tests (e.g., in `tests/checkpointing`) fail. I'd prefer to keep this PR simple, and add a TODO for introducing epoch_end methods in BoringModel in a different PR

File: tests/models/test_cpu.py (hunk 14 above)
Comment (gianscarpe) [reply tests/models/test_cpu.py:234]: Regarding the parameters, I proposed a solution that doesn't require input parameters, while still defining parameters instead of using them as "magic constants". I believe this solution should be clean enough, maybe @Borda can help us on this matter.

File: tests/models/test_cpu.py (hunk 13 above)
Comment (gianscarpe) [reply tests/models/test_cpu.py:104]: I see your point, sounds good to me :)

File: tests/models/test_cpu.py (hunk 15 above)
Comment (rohitgr7) [reply tests/models/test_cpu.py:265]: ok, I see, boring model already has `_epoch_end` methods overridden. It's ok then. Or you can log it in val/test_step, still return the dict just for the sake of the epoch_end methods already defined in the inherited BoringModel.

TODO is not required BTW.

File: tests/base/boring_model.py (hunk 1 above)
Comment (gianscarpe) [reply tests/base/boring_model.py]: Hi @Borda, I chose to keep only in and out features as hparams to build the state of BoringModel. Other parameters are simply specified in __init__ (e.g., `optimizer_name`, ...) to avoid magic numbers and strings. Does it sound good to you?

**Comment by gianscarpe** [comment]
Hi @Borda , @rohitgr7 I made the suggested changes and th PR is ready for review! Thanks :)

File: tests/models/test_cpu.py (hunk 15 above)
Comment (tchaton) [reply tests/models/test_cpu.py:265]: I would prefer have no logging by default in the BoringModel.

Comment (gianscarpe) [reply tests/models/test_cpu.py:265]: @tchaton I agree, in fact I removed logging. You can check the most recent commit for this, in the mean time I resolved this conversation

Cheers! :)

File: tests/models/test_cpu.py [hunk 17]
```diff
@@ -21,16 +21,14 @@
 import tests.base.develop_pipelines as tpipes
 import tests.base.develop_utils as tutils
 from pytorch_lightning import Trainer
-from pytorch_lightning.callbacks import EarlyStopping
-from pytorch_lightning.callbacks import ModelCheckpoint
+from pytorch_lightning.callbacks import Callback, EarlyStopping, ModelCheckpoint
 from pytorch_lightning.core.step_result import TrainResult
```
Comment (rohitgr7) [inline tests/models/test_cpu.py:25]:
```suggestion
```

File: tests/models/test_cpu.py [hunk 18]
```diff
@@ -186,7 +195,29 @@ def test_default_logger_callbacks_cpu_model(tmpdir):
...
-    model = EvalModelTemplate()
+    class ModelTrainValTest(BoringModel):
+        def __init__(self, *args, **kwargs):
+            super().__init__(*args, **kwargs)
+
+        def validation_step(self, batch, batch_idx):
+            output = self.layer(batch)
+            loss = self.loss(batch, output)
+            return {"x": loss}
+
```
Comment (rohitgr7) [inline tests/models/test_cpu.py:206]:
```suggestion
```
same as what we have in BoringModel.

File: tests/models/test_cpu.py [hunk 19]
```diff
@@ -186,7 +195,29 @@ def test_default_logger_callbacks_cpu_model(tmpdir):
...
+        def validation_epoch_end(self, outputs) -> None:
+            val_loss = torch.stack([x["x"] for x in outputs]).mean()
+            self.log('val_loss', val_loss)
+
+        def test_step(self, batch, batch_idx):
+            output = self.layer(batch)
+            loss = self.loss(batch, output)
+            return {"y": loss}
+
```
Comment (rohitgr7) [inline tests/models/test_cpu.py:215]:
```suggestion
```
same here.

File: tests/models/test_cpu.py [hunk 20]
```diff
@@ -202,22 +233,38 @@ def test_running_test_after_fitting(tmpdir):
...
+            super().__init__(*args, **kwargs)
+
+        def val_loader(self):
+            pass
+
+        def test_step(self, batch, batch_idx):
+            output = self.layer(batch)
+            loss = self.loss(batch, output)
+            return {"y": loss}
+
```
Comment (rohitgr7) [inline tests/models/test_cpu.py:262]:
```suggestion
```
same here.

File: tests/models/test_cpu.py [hunk 21]
```diff
@@ -427,85 +378,8 @@ def train_dataloader(self):
...
-    hparams.update(
-        batch_size=batch_size,
-        in_features=truncated_bptt_steps,
-        hidden_dim=truncated_bptt_steps,
-        out_features=truncated_bptt_steps
-    )
-
-    model = BpttTestModel(**hparams)
+    model = BpttTestModel(batch_size=batch_size,
+                          in_features=truncated_bptt_steps, out_features=truncated_bptt_steps)
```
Comment (rohitgr7) [inline tests/models/test_cpu.py:382]:
```suggestion
    model = BpttTestModel(
        batch_size=batch_size, in_features=truncated_bptt_steps, out_features=truncated_bptt_steps
    )
```

File: tests/models/test_cpu.py [hunk 22]
```diff
@@ -75,57 +70,68 @@ def test_cpu_slurm_save_load(tmpdir):
     # new logger file to get meta
     logger = tutils.get_default_logger(tmpdir, version=version)
 
+    model = BoringModel()
+
+    class _StartCallback(Callback):
+        # set the epoch start hook so we can predict before the model does the full training
+        def on_epoch_start(self, trainer, model):
```
Comment (rohitgr7) [inline tests/models/test_cpu.py:77]:
```suggestion
        def on_train_epoch_start(self, trainer, model):
```
some changes will be made to `on_epoch_start` in the future so let's change this too.

File: tests/base/develop_pipelines.py [hunk 23]
```diff
@@ -39,59 +43,64 @@ def run_model_test_without_loggers(trainer_options, model, min_acc: float = 0.50
...
 
     if trainer.use_ddp:
         # on hpc this would work fine... but need to hack it for the purpose of the test
         trainer.model = pretrained_model
         trainer.optimizers, trainer.lr_schedulers = pretrained_model.configure_optimizers()
 
 
-def run_model_test(trainer_options, model, on_gpu: bool = True, version=None, with_hpc: bool = True):
+def run_model_test(trainer_options, model, on_gpu: bool = True, version=None,
+                   with_hpc: bool = True, min_acc: float = 0.25):
```
Comment (rohitgr7) [inline tests/base/develop_pipelines.py:55]:
```suggestion
def run_model_test(
    trainer_options, model, on_gpu: bool = True, version=None, with_hpc: bool = True, min_acc: float = 0.25
):
```

**Review by rohitgr7 — COMMENTED** [review]
Just some more suggestions. Else LGTM.

**Comment by gianscarpe** [comment]
@Borda Hi, I merged master branch and solved the issues :)

File: tests/models/test_cpu.py (hunk 18 above)
Comment (gianscarpe) [reply tests/models/test_cpu.py:213]: Thank you! I commit :)

**Comment by gianscarpe** [comment]
@rohitgr7 Thank you for the effort in reviewing my PR! I really appreciated it :)

**Review by rohitgr7 — APPROVED** [review]
LGTM. Nice work!!

File: tests/base/boring_model.py [hunk 24]
```diff
@@ -75,7 +75,12 @@ def training_step(...):
 
         """
         super().__init__()
-        self.layer = torch.nn.Linear(32, 2)
+
+        self.batch_size = 1
+        self.learning_rate = 0.1
+        self.optimizer_name = "SGD"
+        self.layer = torch.nn.Linear(in_features, out_features)
+        self.save_hyperparameters()
```
Comment (Borda) [inline tests/base/boring_model.py:83]: let's move these parameters to init args, as now they won't be saved with `save_hyperparameters`

File: tests/base/develop_pipelines.py (hunk 5 above)
Comment (Borda) [reply tests/base/develop_pipelines.py:131]: any change here?

**Review by Borda — COMMENTED** [review]
lest drop all " replacement it makes the PR too large and it is unrequested change, also lets the import formatting do as part of #4805

File: tests/base/boring_model.py (hunk 24 above)
Comment (gianscarpe) [reply tests/base/boring_model.py]: Sure! I moved parameters to __init__ :100:

**Comment by gianscarpe** [comment]
@Borda I drop the unnecessary changes (' and " and newlines) to reduce PR size! :) In `develop_pipelines` I already simplified `run_prediction` for `BoringModel` as you requested :)

**Comment by Borda** [comment]
> @Borda I drop the unnecessary changes (' and " and newlines) to reduce PR size! :)

eventy=ualy we can do it but lest do it in separate PR and most like in multiple smaller steps, and then we can merge it with fixing imports...

**Comment by gianscarpe** [comment]
Sure @Borda ! I reduced the PR, now it should introduce only the necessary changes and deprecate BoringModel in `test_cpu` and `test_gpu`. Is it right?

File: tests/base/boring_model.py [hunk 25]
```diff
@@ -55,8 +56,9 @@ def __len__(self):
 
 
 class BoringModel(LightningModule):
-
-    def __init__(self):
+    def __init__(self, in_features: int = 32, out_features: int = 2,
+                 optimizer_name: str = 'SGD', learning_rate: float = 0.1,
+                 batch_size:int = 1):
```
Comment (rohitgr7) [inline tests/base/boring_model.py:61]: I'd still suggest not adding any arguments or any other assignments in BoringModel. Recently I added `example_input_array` which is a lightning thing but William suggested to remove that and instead override Boring Model to keep BoringModel as it is.

Comment (gianscarpe) [reply tests/base/boring_model.py:61]: Thank you @rohitgr7! @Borda can you help me on this? What do you suggest?

**Review by tchaton — APPROVED** [review]
LGTM !

File: tests/base/boring_model.py [hunk 26]
```diff
@@ -117,15 +121,16 @@ def test_epoch_end(self, outputs) -> None:
...
     def configure_optimizers(self):
-        optimizer = torch.optim.SGD(self.layer.parameters(), lr=0.1)
+        optimizer_class = getattr(torch.optim, self.optimizer_name)
+        optimizer = optimizer_class(self.parameters(), lr=self.learning_rate)
         lr_scheduler = torch.optim.lr_scheduler.StepLR(optimizer, step_size=1)
         return [optimizer], [lr_scheduler]
 
     def train_dataloader(self):
-        return torch.utils.data.DataLoader(RandomDataset(32, 64))
+        return torch.utils.data.DataLoader(RandomDataset(32, 64), batch_size=self.batch_size)
```
Comment (tchaton) [inline tests/base/boring_model.py:130]: Can you use in_features there too. If I change it, it will break

Comment (gianscarpe) [reply tests/base/boring_model.py]: Thanks @tchaton! I totally agree! Done the requested change :+1:

**Comment by awaelchli** [comment]
Isn't the point of BoringModel that it is boring and simple with the most minimal functional logic? I think I saw @williamFalcon emphasizing this several times so just make sure it's good with him :)

**Review by williamFalcon — DISMISSED** [review]
yes please. keep the boring model BORING :).

This model does NOT need to ever be changed.

Tests should create a subclass of this and implement whatever custom functionality they need

**Comment by gianscarpe** [comment]
> yes please. keep the boring model BORING :).
> 
> This model does NOT need to ever be changed.
> 
> Tests should create a subclass of this and implement whatever custom functionality they need

Thank you @williamFalcon, I reverted BoringModel and inherited it in specific tests of `test_cpu` for custom features :+1:  Maybe we can discuss this policy a bit, I think this approach could lead to copy-and-paste tests if specific functionalities are needed in different tests. Maybe a more structured subclass hierarchy could help in the future. Just an idea, and btw great job with pytorch-lightning! I'm here to learn, so thank you :)

**Comment by gianscarpe** [comment]
I ping the discussion, in case more changes are required @williamFalcon @Borda :)

File: tests/models/test_cpu.py [hunk 27]
```diff
@@ -218,12 +240,23 @@ def test_running_test_after_fitting(tmpdir):
...
-    tutils.assert_ok_model_acc(trainer, thr=0.5)
+    tutils.assert_ok_model_acc(trainer, key='test_loss', thr=0.5)
 
 
 def test_running_test_no_val(tmpdir):
     """Verify `test()` works on a model with no `val_loader`."""
-    model = EvalModelTemplate()
+    class ModelTrainTest(BoringModel):
+        def __init__(self, *args, **kwargs):
+            super().__init__(*args, **kwargs)
```
Comment (carmocca) [inline tests/models/test_cpu.py:250]: Don't need this

```suggestion
```

File: tests/models/test_cpu.py [hunk 28]
```diff
@@ -192,7 +202,19 @@ def test_default_logger_callbacks_cpu_model(tmpdir):
 
 def test_running_test_after_fitting(tmpdir):
     """Verify test() on fitted model."""
-    model = EvalModelTemplate()
+    class ModelTrainValTest(BoringModel):
+        def __init__(self, *args, **kwargs):
+            super().__init__(*args, **kwargs)
```
Comment (carmocca) [inline tests/models/test_cpu.py:207]: same

```suggestion
```

File: tests/models/test_cpu.py [hunk 29]
```diff
@@ -104,20 +100,31 @@ def assert_pred_same():
 @pytest.mark.parametrize("enable_pl_optimizer", [False, True])
 def test_early_stopping_cpu_model(enable_pl_optimizer, tmpdir):
     """Test each of the trainer options."""
-    stopping = EarlyStopping(monitor='early_stop_on', min_delta=0.1)
+    class ModelTrainVal(BoringModel):
+        def __init__(self, *args, **kwargs):
+            super().__init__(*args, **kwargs)
```
Comment (carmocca) [inline tests/models/test_cpu.py:105]: same :)

```suggestion
```

File: tests/models/test_cpu.py [hunk 30]
```diff
@@ -218,12 +240,23 @@ def test_running_test_after_fitting(tmpdir):
...
 
 def test_running_test_no_val(tmpdir):
     """Verify `test()` works on a model with no `val_loader`."""
-    model = EvalModelTemplate()
+    class ModelTrainTest(BoringModel):
+        def __init__(self, *args, **kwargs):
+            super().__init__(*args, **kwargs)
+
+        def val_loader(self):
+            pass
```
Comment (carmocca) [inline tests/models/test_cpu.py:253]: Is this test relevant anymore? BoringModel does not have a `val_loader` attribute. Is it supposed to be `val_dataloader`?

File: tests/models/test_cpu.py (hunk 27 above)
Comment (gianscarpe) [reply tests/models/test_cpu.py:250]: Thanks! Make it cleaner :)

**Comment by gianscarpe** [comment]
Hi @Borda @williamFalcon, can you check this PR? Many thanks :)

File: tests/base/develop_pipelines.py [hunk 31]
```diff
@@ -56,7 +61,8 @@ def run_model_test(trainer_options, model, on_gpu: bool = True, version=None, wi
     logger = get_default_logger(save_dir, version=version)
     trainer_options.update(logger=logger)
 
-    if 'checkpoint_callback' not in trainer_options:
+    # TODO: DEPRECATED option
+    if "checkpoint_callback" not in trainer_options:
         trainer_options.update(checkpoint_callback=True)
```
Comment (Borda) [inline tests/base/develop_pipelines.py:66]: shall we remove it?

File: tests/base/develop_pipelines.py [hunk 32]
```diff
@@ -119,3 +128,14 @@ def run_prediction(dataloader, trained_model, dp=False, min_acc=0.50):
         acc = acc.item()
 
     assert acc >= min_acc, f"This model is expected to get > {min_acc} in test set (it got {acc})"
+
+
+@run_prediction.register(BoringModel)
```
Comment (Borda) [inline tests/base/develop_pipelines.py:133]: I am still not very convinced about this wrapping, I see the beauty but it is not trivial for new users...

Comment (gianscarpe) [reply tests/base/develop_pipelines.py:133]: Hi @borda and thanks for the review! I see your point, this could be hard to follow for someone new to python.Sadly, python doesn't have polimorphysm. If we don't want to add extra methods to `boring model`, the only way is to dispatch `run_predictions` calls here, distinguishing per classes and subclasses (a lot of tests inherit BoringModel). This is what`singledispatch` is for. Maybe i can add some comments to explain the approach, does it sound good to you? 
Otherwise, we need to use nested if for dispatching, but then the code loses in clarity (at least for me) . In any case no problem, let me know and I'll fix this :) 
Many thanks!

Comment (Borda) [reply tests/base/develop_pipelines.py:133]: lets see what others think... @PyTorchLightning/core-contributors 
for me, we shall keep the contribution barrier as low as possible

File: tests/base/develop_pipelines.py (hunk 31 above)
Comment (gianscarpe) [reply tests/base/develop_pipelines.py:66]: I'll fix this in order to make it right as part of this PR :)

Comment (gianscarpe) [reply tests/base/develop_pipelines.py:66]: Hi @Borda, I think we can remove it as it's not working anymore. We can add this feature in the feature :)

**Comment by gianscarpe** [comment]
Hi @Borda, I think you're right about this. I commit I possible solution. I'm not a fun, but I think it's easier to to understand. let me know :)

File: tests/models/test_cpu.py (hunk 30 above)
Comment (carmocca) [reply tests/models/test_cpu.py:247]: Doesn't seem fixed :thinking: 

Also update the docstring!

Comment (gianscarpe) [reply tests/models/test_cpu.py]: @carmocca fixed it! Thank you :)

**Review by s-rog — APPROVED** [review]

**Review by carmocca — APPROVED** [review]

**Comment by gianscarpe** [comment]
Ping for review @Borda :)

**Review by Borda — APPROVED** [review]
lgtm

**Comment by Borda** [comment]
@williamFalcon @tchaton mind re-review?

**Comment by gianscarpe** [comment]
Hi @tchaton! Does this sound good to you as well? Thanks! :)

**Comment by Borda** [comment]
@gianscarpe sorry for bordering, but just thinking that this maybe make better sense to land in feature 1.2
what do you think, I can help with the change so no need to closing or redo your work...

File: tests/metrics/utils.py [hunk 33]
```diff
@@ -11,6 +11,11 @@
 
 from pytorch_lightning.metrics import Metric
 
+try:
+    set_start_method("spawn")
+except RuntimeError:
+    pass
+
```
Comment (Borda) [inline tests/metrics/utils.py:18]: why this is moved here?

File: tests/metrics/regression/test_ssim.py [hunk 34]
```diff
@@ -53,9 +53,7 @@ def _sk_metric(preds, target, data_range, multichannel):
 class TestSSIM(MetricTester):
     atol = 6e-5
 
-    # TODO: for some reason this test hangs with ddp=True
```
Comment (Borda) [inline tests/metrics/regression/test_ssim.py:56]: is this resolved? if not pls keep the note...

**Comment by Borda** [comment]
@gianscarpe feat-1.2 shall be fixed now, pls update your PR ;]

File: pytorch_lightning/trainer/supporters.py [hunk 35]
```diff
@@ -253,34 +251,19 @@ class CombinedDataset(object):
     """
     Combine multiple datasets and compute their statistics
     """
-    COMPUTE_FUNCS = {'min_size': min, 'max_size_cycle': max}
-
-    def __init__(self, datasets: Union[Sequence, Mapping], mode: str = 'min_size'):
+    def __init__(self, datasets: Union[Sequence, Mapping], mode: str):
```
Comment (Borda) [inline pytorch_lightning/trainer/supporters.py:254]: seems you are reverting #5352

**Review by Borda — CHANGES_REQUESTED** [review]

File: pytorch_lightning/trainer/supporters.py (hunk 35 above)
Comment (gianscarpe) [reply pytorch_lightning/trainer/supporters.py]: Solve!

**Review by Borda — APPROVED** [review]

**Comment by Borda** [comment]
@tchaton mind re-review? ^^

**Review by tchaton — APPROVED** [review]
